kuka_iiqka_eac_driver/src/hardware_interface.cpp

```diff
@@ -228,135 +181,74 @@ CallbackReturn KukaEACHardwareInterface::on_configure(const rclcpp_lifecycle::St
     info_.hardware_parameters.at("timeframe_ms").c_str());
-#endif
+
   return CallbackReturn::SUCCESS;
 }
 
 CallbackReturn KukaEACHardwareInterface::on_activate(const rclcpp_lifecycle::State &)
 {
-  RCLCPP_INFO(rclcpp::get_logger("KukaEACHardwareInterface"), "Connecting to robot . . .");
-  // Reset timeout to catch first tick message
-  receive_timeout_ = std::chrono::milliseconds(100);
-  control_signal_ext_.control_signal.stop_ipo = false;
-#ifdef NON_MOCK_SETUP
-  if (context_ != nullptr)
+  kuka::external::control::OperationStatus create_event_observer =
+    robot_ptr_->RegisterEventHandler(std::make_unique<KukaEACEventObserver>(this));
+  if (create_event_observer.return_code == kuka::external::control::ReturnCode::ERROR)
   {
-    context_->TryCancel();
+    RCLCPP_ERROR(
+      rclcpp::get_logger("KukaEACHardwareInterface"),
+      "Creating event observer failed, error message: %s", create_event_observer.message);
   }
-#endif
-  if (observe_thread_.joinable())
+
+  kuka::external::control::OperationStatus start_control = robot_ptr_->StartControlling(
+    static_cast<kuka::external::control::ControlMode>(hw_control_mode_command_));
+  if (start_control.return_code == kuka::external::control::ReturnCode::ERROR)
   {
-    observe_thread_.join();
+    RCLCPP_ERROR(
+      rclcpp::get_logger("KukaEACHardwareInterface"),
+      "Starting external control failed, error message: %s", start_control.message);
+    return CallbackReturn::FAILURE;
   }
 
-#ifdef NON_MOCK_SETUP
-  observe_thread_ = std::thread(&KukaEACHardwareInterface::ObserveControl, this);
-
-  OpenControlChannelRequest request;
-  OpenControlChannelResponse response;
-  grpc::ClientContext context;
+  prev_control_mode_ = static_cast<int>(hw_control_mode_command_);
 
-  request.set_ip_address(info_.hardware_parameters.at("client_ip"));
-  request.set_timeout(5000);
-  request.set_cycle_time(4);
-  request.set_external_control_mode(
-    kuka::motion::external::ExternalControlMode(hw_control_mode_command_));
   RCLCPP_INFO(
-    rclcpp::get_logger("KukaEACHardwareInterface"), "Starting control in %s",
-    kuka::motion::external::ExternalControlMode_Name(static_cast<int>(hw_control_mode_command_))
-      .c_str());
-
-  auto ret = stub_->OpenControlChannel(&context, request, &response);
-  if (ret.error_code() != grpc::StatusCode::OK)
-  {
-    RCLCPP_ERROR(rclcpp::get_logger("KukaEACHardwareInterface"), "%s", ret.error_message().c_str());
-    return CallbackReturn::FAILURE;
-  }
-#endif
+    rclcpp::get_logger("KukaEACHardwareInterface"),
+    "External control session started successfully");
 
+  stop_requested_ = false;
   return CallbackReturn::SUCCESS;
 }
 
 CallbackReturn KukaEACHardwareInterface::on_deactivate(const rclcpp_lifecycle::State &)
 {
-  RCLCPP_INFO(rclcpp::get_logger("KukaEACHardwareInterface"), "Deactivating");
+  RCLCPP_INFO(rclcpp::get_logger("KukaEACHardwareInterface"), "Deactivating hardware interface");
```

Comment:
Sometimes this appears multiple times, investigate

Reply:
The hw if does not have a deactivate() function that also triggers the state transition, therefore the callback is called twic (event reaction + robot manager triggering transition)
As only a stop flag is set, it is safer to call it twice, than waiting for the robot manager event

Reply:
fixed with `set_stop_flag()`